### PR TITLE
PPCAnalyst: track CR outputs individually

### PR DESCRIPTION
**Pull request by FioraAeterna**
More generic and cleaner and correct than what we have now. Not used for
anything yet.

I figure I might as well get in some of my diffs lying around since I'll be breaking from dolphin dev in a little bit (~work~)

### CODE REVIEW

**Comment by FioraAeterna** [comment]
Please don't merge this before https://github.com/FioraAeterna/dolphin/compare/discardunused since it'll conflict heavily, I mainly wanted to get this out there.

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_Tables.cpp [hunk 1]
```diff
@@ -147,14 +147,14 @@ static GekkoOPTemplate table19[] =
...
-	{417, Interpreter::crorc,       {"crorc",  OPTYPE_CR, FL_EVIL, 1, 0, 0, 0}},
-	{193, Interpreter::crxor,       {"crxor",  OPTYPE_CR, FL_EVIL, 1, 0, 0, 0}},
+	{257, Interpreter::crand,       {"crand",  OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
+	{129, Interpreter::crandc,      {"crandc", OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
+	{289, Interpreter::creqv,       {"creqv",  OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
+	{225, Interpreter::crnand,      {"crnand", OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
+	{33,  Interpreter::crnor,       {"crnor",  OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
+	{449, Interpreter::cror,        {"cror",   OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
+	{417, Interpreter::crorc,       {"crorc",  OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
+	{193, Interpreter::crxor,       {"crxor",  OPTYPE_CR, FL_EVIL | FL_SET_CRn, 1, 0, 0, 0}},
```
Comment (magumagu) [inline Source/Core/Core/PowerPC/Interpreter/Interpreter_Tables.cpp]: Do these instructions still need FL_EVIL given that we have FL_SET_CRn?  Or do we also need FL_IN_CR or something like that?  (It seems like a bad idea to have our generic "don't optimize this" flag attached to a commonly used instruction.)

Comment (FioraAeterna) [inline Source/Core/Core/PowerPC/Interpreter/Interpreter_Tables.cpp]: Yeah, I don't know about FL_EVIL, but that sounds reasonable. We are starting to run out of flags though; I guess we could raise the flags to 64-bit from 32-bit.

File: Source/Core/Common/BitSet.h [hunk 2]
```diff
@@ -34,8 +40,10 @@ static inline int LeastSignificantSetBit(u64 val)
 	return (int)index;
 }
 #else
+static inline int CountSetBits(u8 val) { return __builtin_popcount(val); }
 static inline int CountSetBits(u32 val) { return __builtin_popcount(val); }
 static inline int CountSetBits(u64 val) { return __builtin_popcountll(val); }
+static inline int LeastSignificantSetBit(u8 val) { return __builtin_ctz(val); }
 static inline int LeastSignificantSetBit(u32 val) { return __builtin_ctz(val); }
 static inline int LeastSignificantSetBit(u64 val) { return __builtin_ctzll(val); }
 #endif
```
Comment (magumagu) [inline Source/Core/Common/BitSet.h]: Might as well add 16-bit variants while you're here.

File: Source/Core/Core/PowerPC/PPCAnalyst.cpp [hunk 3]
```diff
@@ -476,30 +476,28 @@ void PPCAnalyzer::ReorderInstructions(u32 instructions, CodeOp *code)
...
-		code->outputCR1 = true;
-	else
-		code->outputCR1 = (opinfo->flags & FL_SET_CR1) ? true : false;
+	code->outputCR = BitSet8(0);
+	if ((opinfo->flags & FL_RC_BIT && (code->inst.hex & 1)) || (opinfo->flags & FL_SET_CR0))
+		code->outputCR[0] = true;
+	if ((opinfo->flags & FL_RC_BIT_F && (code->inst.hex & 1)) || (opinfo->flags & FL_SET_CR1))
+		code->outputCR[1] = true;
+	// mtcrf: this one is special and can modify multiple CR regs
+	if (code->inst.OPCD == 31 && code->inst.SUBOP10 == 144)
```
Comment (magumagu) [inline Source/Core/Core/PowerPC/PPCAnalyst.cpp]: Maybe add a helper function isMtcrf?

Comment (FioraAeterna) [inline Source/Core/Core/PowerPC/PPCAnalyst.cpp]: hmm, not a bad idea, though if we do that we have another ~dozen instructions we'll need to add helper functions for to be consistent, I think.

**Comment by skidau** [comment]
@dolphin-emu-bot allowmerge

**Comment by comex** [comment]
@FioraAeterna Isn't this ready?  Please update it or I'm going to close it. :)

**Pull request closed without merging**